src/Microsoft.ML.FastTree/TreeEnsemble/TreeEnsembleCombiner.cs

```diff
@@ -65,9 +65,9 @@ IPredictor IModelCombiner.CombineModels(IEnumerable<IPredictor> models)
                 foreach (var t in tree.TrainedEnsemble.Trees)
                 {
                     var bytes = new byte[t.SizeInBytes()];
-                    int position = -1;
```

Comment:
Interesting. Can you please explain this issue a bit more? Why does this happen in x64 but not in x86? This is managed memory. Why is it corrupting the unamanaged heap?

Reply:
Yes, this is index out of range issue that corrupts memory but this one is index was out if range in former (start to use byte array from index -1)...
This memory corrupted (byte array) is allocated in managed but used as unmanaged (from fixed section in C# and pointer) like below:
[https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.FastTree/TreeEnsemble/InternalRegressionTree.cs#L101](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.FastTree/TreeEnsemble/InternalRegressionTree.cs#L101)
[https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.FastTree/Utils/ToByteArrayExtensions.cs#L113](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.FastTree/Utils/ToByteArrayExtensions.cs#L113)
when position is -1, the pointer ((int*)(pBuffer + position)) is accessing memory it should not.


I'm still not sure why this issue not repro in x64. In theory this can also corrupt x64 memory.

---
In reply to: [419011127](https://github.com/dotnet/machinelearning/pull/5081#discussion_r419011127) [](ancestors = 419011127)

Reply:
SZArrays on x64 have a 4-byte padding between the length and the start of the data.

https://github.com/dotnet/runtime/blob/200b197559fc49f91b8db3f20fd6df6b5660d7bf/src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs#L306-L321

Reply:
I suspected this was the issue from reading above, but special thanks to @stephentoub for finding the actual code that proves it 👍

Reply:
@eerhardt also confirmed this in email. 
@sharwell, @stephentoub and @eerhardt Thanks for shedding more light on the issue.